### ISSUE TITLE
Connection pooling appears broken when using AccessTokens with SQL Azure

### ISSUE THREAD

**Issue by MatthewSteeples** [AzureAD/azure-activedirectory-library-for-dotnet#1462]
**Which Version of ADAL are you using ?**
ADAL 4.4.1 (latest from NuGet)

**Which platform has the issue?**
net472 (Azure App Service)

**What authentication flow has the issue?**
`Task<AuthenticationResult> AcquireTokenAsync(string resource, IClientAssertionCertificate clientCertificate)`

**Repro**
We use this method to populate the AccessToken property of a SqlConnection object to talk to SQL Azure. This was working fine in 3.19.8, but when we updated the package (with no other code changes) to 4.4.1 we found that SQL connections were no longer pooled (so we quickly went up to using over 5000 connections when our baseline is less than 100).

```csharp
var clientId = "GuidFromAzureAD";
var cert = new X509Certificate2("AzureADCert");
ClientAssertionCertificate certCred = new ClientAssertionCertificate(clientId, cert);
AuthenticationContext authContext = new AuthenticationContext(authority, TokenCache.DefaultShared);
var token = await authContext.AcquireTokenAsync(resourceId, certCred);
var sqlConnection = new SqlConnection("SqlAzureConnectionString");
sqlConnection.Token = token.AccessToken;
```

**Expected behavior**
Connection pooling works as it did in 3.19.8

**Actual behavior**
Number of connections created increases dramatically

**Possible Solution**
Haven't looked yet, but the only resolution we came up with was to roll back to 3.19.8

**Comment by jmprieur** [comment]
@MatthewSteeples : I understand what you write, but don't understand how this is related to ADAL.NET / authentication as ADAL.NET does not manage SQL connections.

Can you please tell me if you observe something different from the point of view of tokens between ADAL.NET 3.19.8 and ADAL.NET 4.4.1 ? For instance, do I understand that, when you are getting the token in ADA.NET 3.19.8, and assign it as the `AccessToken` of the sqlConnection, the token the exact same as in a previous call (unless it has to be refreshed) whereas now it's refreshed each time / not cached?

**Comment by MatthewSteeples** [comment]
@jmprieur Thanks for your quick response. I enabled logging to try and narrow it down based on your questions, but it doesn't make much sense. Both versions retrieve the token from the cache (so do not communicate with AD) after the first request and both versions return the exact same AccessToken that is then passed in to SQL Server.

That is where the similarities stop, because the 3.19.8 library then re-uses the connection, but the 4.4.1 library creates a new request per iteration. I literally changed nothing else apart from the versions of the libraries between the 2 tests.

I'm not sure how this library is at fault, but it is definitely making a difference somewhere. I can reproduce it reliably in our solution, so my next step is to create a minimal-reproduction that I can share with you (and I'll provide credentials to a SQL Database too) so you can see it for yourself.

**Comment by MatthewSteeples** [comment]
OK, this is weird. I've got a minimal repro (just need to jump through the hoops of making a cert etc so I can share the credentials, so I'll attache that soon) but I've noticed that the smallest change which should be irrelevant makes a big difference: just the act of calling AcquireAccessTokenAsync is enough to make it not re-use the connection, even though the token being returned is identical.

If I generate the accesstoken each time I go through the loop, the connection is not re-used. If I generate the accesstoken outside the loop (so it's only called once) the connection is re-used.

Full code to follow

**Comment by jmprieur** [comment]
Thanks @MatthewSteeples for investigating and proposing to share a minimal repro, this will be very helpful. 
So I understand that there is a weird interaction between the fact of calling AcquireTokenAsync, and the internals of the SqlConnection pool. Maybe something related to Tasks? (Amazing, side effect, BTW!)

heads-up to : @MarkZuber @henrik-me @bgavrilMS

**Comment by MatthewSteeples** [comment]
@jmprieur Repro attached. The program runs against a DB in my subscription (which I can grant you access to if needed, it's empty). The AD application is a member of db_owner for this database. All keys and secrets are pre-configured. The program will print out the results of query that shows how many connections are open for this Application Name.

To see the alternative behaviour, move the token generation (line 29) out of the loop (line 24) and re-run. You'll see that the number of connections does not increment.

[SqlAdBug.zip](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/files/2676288/SqlAdBug.zip)

**Comment by MatthewSteeples** [comment]
The behaviour that this app shows does not happen if you change the version of ADAL installed to be 3.19.8 (attached to save the inconvenience of fiddling round with NuGet)

[SqlAdBug3198.zip](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/files/2676294/SqlAdBug3198.zip)

**Comment by MarkZuber** [comment]
I've verified that we get EXACTLY the same access token on each call through the loop.
However, it looks like something in the SQL layer is looking at REFERENCE equality and not VALUE equality when determining if a new secure connection in the pool needs to be made.  

This doesn't (yet) explain why ADAL 3.19.8 doesn't exhibit this behavior but it might at least provide you a work around or provide more context on what's going on.

Investigation steps:
The code below caches the original value of the access token in a string.  
We then continue to get a new token each time and compare the values.  
The values compare equal every time.  
If I send in "originalAccessToken" (same VALUE, same REFERENCE) then the connection pool does not increase.  
If I send in "token.AccessToken" (same VALUE, different REFERENCE) then the connection pool DOES increase.

            AuthenticationContext authContext = new AuthenticationContext(authority, TokenCache.DefaultShared);

            string originalAccessToken = null;

            AuthenticationResult previousResult = null;

            for (int i = 0; i < 1000; i++)
            {
                using (var sqlConnection = new SqlConnection(ConfigurationManager.ConnectionStrings["SqlConnection"].ConnectionString))
                {
                    var token = await authContext.AcquireTokenAsync(resourceId, new ClientCredential(clientId, clientSecret));

                    if (string.IsNullOrEmpty(originalAccessToken))
                    {
                        originalAccessToken = token.AccessToken;
                    }

                    if (previousResult != null)
                    {
                        bool areEqual = string.CompareOrdinal(previousResult.AccessToken, token.AccessToken) == 0;
                        Console.WriteLine($"AreEqual: {areEqual}");
                    }

                    previousResult = token;

                    Console.WriteLine(token.AccessToken.MD5Hash());

                    sqlConnection.AccessToken = originalAccessToken;
                    sqlConnection.Open();

**Comment by MatthewSteeples** [comment]
I've had a look at how SqlConnection is implemented and can't see anything that stands out. It uses GetHashCode from the AccessToken to determine whether the connection has already been pooled or not, but the hash code of the AccessToken returned (like the MD5) is equivalent

I've created a PR which resolves this issue _possibly for only our scenario_. The current behaviour uses a Serialization roundtrip (Serialize to JSON and then Deserialize) to clone the object, whereas the behaviour in the PR clones the object explicitly. This means that the strings are re-used. Obviously this could be a maintenance issue, because any properties added to the classes will need to be added to the clone methods too.

I've had a look at the other places where Serialize and Deserialize are used, but don't think this kind of change applies (and think it's only used in MSAL logic)

**Comment by MatthewSteeples** [comment]
The PR also has the side effect of being quicker (fewer allocations and less CPU time required)

**Comment by MarkZuber** [comment]
Thank you Matthew for your detailed repro steps and analysis.  The effort and level of detail you are giving here is greatly appreciated.  Our team here is looking at this and will get back to you soon.  I don't think we will be issuing any ADAL releases before the end of the calendar year but we should be able to review and give you some feedback.  If it's approved it would be in a release early January 2019.

**Comment by henrik-me** [comment]
This is included in ADAL v 4.5.0
